main/src/vocab/nllb_vocab.rs

```diff
@@ -0,0 +1,238 @@
+    pub language_codes_bytes: HashSet<Vec<u8>>,
+
+    pub special_token_storage: Option<SpecialTokens>,
+}
+
+impl NLLBVocab {
+    /// The beginning of sequence token that was used during pretraining.
+    /// Can be used a sequence classifier token.
+    pub fn bos_value(&self) -> &str {
+        self.special_token_storage
+            .as_ref()
+            .map(|s| s.begin_of_sequence_token.as_str())
+            .unwrap_or("<s>")
+    }
+
+    /// End of sequence token.
+    pub fn eos_value(&self) -> &str {
+        self.special_token_storage
+            .as_ref()
+            .map(|s| s.end_of_sequence_token.as_str())
+            .unwrap_or("</s>")
+    }
+
+    /// Returns the SEP token for M2M100 (`</s>`)
+    pub fn sep_value(&self) -> &str {
+        self.special_token_storage
+            .as_ref()
+            .map(|s| s.end_of_sequence_token.as_str())
+            .unwrap_or("</s>")
+    }
+
+    /// Returns the PAD token for M2M100 (`<pad>`)
+    pub fn pad_value(&self) -> &str {
+        self.special_token_storage
+            .as_ref()
+            .map(|s| s.padding_token.as_str())
+            .unwrap_or("<pad>")
+    }
+}
+
+impl Vocab for NLLBVocab {
+    fn unknown_value() -> &'static str {
+        "<unk>"
+    }
+
+    fn get_unknown_value(&self) -> &str {
+        self.special_token_storage
+            .as_ref()
+            .map(|s| s.unknown_token.as_str())
+            .unwrap_or("<unk>")
+    }
+
+    fn values(&self) -> &HashMap<String, i64> {
+        &self.values
+    }
+
+    fn indices(&self) -> &HashMap<i64, String> {
+        &self.indices
+    }
+
+    fn special_values(&self) -> &HashMap<String, i64> {
+        &self.special_values
+    }
+
+    fn special_indices(&self) -> &HashMap<i64, String> {
+        &self.special_indices
+    }
+
+    fn from_file<V: AsRef<Path>, S: AsRef<Path>>(
+        vocab: V,
+        special: Option<S>,
+    ) -> Result<Self, TokenizerError> {
+        let special_token_storage: Option<SpecialTokens> = special
+            .map(|p| {
+                File::open(&p)
+                    .context(IOSnafu { path: p.as_ref() })
+                    .and_then(|r| serde_json::from_reader(r).context(JsonDeserializeSnafu))
+            })
+            .transpose()?;
+
+        let mut tokenizer: Tokenizer = File::open(&vocab)
+            .context(IOSnafu {
+                path: vocab.as_ref(),
+            })
+            .map(BufReader::new)
+            .and_then(|r| serde_json::from_reader(r).context(JsonDeserializeSnafu))?;
+
+        let mut special_values = HashMap::with_capacity(FAIRSEQ_LANGUAGE_CODES.len() + 10);
+
+        let values = &mut tokenizer.model.vocab;
+        for language_code in FAIRSEQ_LANGUAGE_CODES.iter() {
+            let language_code = if language_code.len() == 3 {
+                format!(">>{language_code}<<")
+            } else {
+                return VocabularyValidationSnafu {
```

Comment:
I am wondering how this passes given all language codes are of length 8 and the check above passes if the length is 3?

Reply:
This code is inconsistent and it's not working, sorry for this:(

I commited it before leaving the office, I'll fix it tomorrow.

Reply:
Well, I fixed this and some other places. And it works 🎉